packages/template-retail-react-app/app/contexts/index.js

```diff
@@ -31,19 +33,160 @@ import PropTypes from 'prop-types'
-export const CategoriesProvider = ({categories: initialCategories = {}, children}) => {
-    const [categories, setCategories] = useState(initialCategories)
+export const CategoriesProvider = ({treeRoot = {}, children, testData = {}}) => {
+    const itemsKey = 'categories'
+    const DEFAULT_ROOT_CATEGORY = 'root'
+    const LOCAL_STORAGE_PREFIX = `pwa-kit-cat-`
+
+    const api = useCommerceAPI()
+    const [root, setRoot] = useState({
+        // map over the server-provided cat
+        ...treeRoot[DEFAULT_ROOT_CATEGORY],
+        [itemsKey]: treeRoot[DEFAULT_ROOT_CATEGORY]?.[itemsKey]?.map((item) => ({
+            ...item,
+            loaded: false
+        }))
+    })
+    const [queue, setQueue] = useState({})
+
+    // iterates through each deep nested object and if finds object that has prop and value specified in objToFindBy
+    // argument, it replaces the current object with replacementObj, stops recursive walk and returns the whole tree.
+    // If none is found, it returns false.
+
+    // copied from https://github.com/brojd/obj-traverse/blob/master/src/findAndModifyFirst/findAndModifyFirst.js
+    const findAndModifyFirst = (tree, childrenKey, objToFindBy, replacementObj) => {
+        let treeToReturn = tree
+        let findSuccess = false
+        let modifiedObj = false
+        const findKeys = Object.keys(objToFindBy)
+        findKeys.forEach((key) => {
+            isEqual(tree[key], objToFindBy[key]) ? (findSuccess = true) : (findSuccess = false)
+        })
+        if (findSuccess) {
+            for (let prop in tree) {
+                delete tree[prop]
+            }
+            for (let prop in replacementObj) {
+                tree[prop] = replacementObj[prop]
+            }
+            return tree
+        }
+        const findInChildren = (obj, childrenKey, objToFindBy, replacementObj) => {
+            if (Object.prototype.hasOwnProperty.call(obj, childrenKey)) {
+                for (let i = 0; i < obj[childrenKey].length; i++) {
+                    findKeys.forEach((key) => {
+                        isEqual(obj[childrenKey][i][key], objToFindBy[key])
+                            ? (findSuccess = true)
+                            : (findSuccess = false)
+                    })
+                    if (findSuccess) {
+                        obj[childrenKey][i] = replacementObj
+                        modifiedObj = true
+                        break
+                    }
+                }
+                if (!findSuccess) {
+                    obj[childrenKey].forEach((child) =>
+                        findInChildren(child, childrenKey, objToFindBy, replacementObj)
+                    )
+                }
+            }
+            return obj
+        }
+        findInChildren(tree, childrenKey, objToFindBy, replacementObj)
+        return modifiedObj ? treeToReturn : false
+    }
+
+    const fetchCategoryNode = async (id, depth = 1) => {
+        const STALE_TIME = 10000
```

Comment:
This is an important constant and should probably be lifted somewhere. Also, comments with units are helpful for "magic" constants.
```suggestion
        const STALE_TIME = 10000 // 10 seconds
```